### PR TITLE
[MM-133] Added a feature of restricting bot messages to private channels and DM/GMs

### PR DESCRIPTION
**Pull request by ayusht2810**
#### Summary
- Added a feature to restrict bot messages to private channels and DM/GMs through plugin config.
- Added a feature to restrict bot messages to a particular channel via the slash command.
- Added a feature to list all channel preferences.

#### Ticket Link
#79 

#### Screenshots
image
image
image
image

#### What to test?
- Restricting bot posting meeting links in the channels when changing settings through plugin config or through slash commands.
- View list of channels with updated preferences.
- Disabling of `settings` slash command when posting meeting links is restricted.

### CODE REVIEW

File: README.md
Comment (raghavaggarwal2308) [inline README.md:221]:
```suggestion
- `/zoom connect` :- To connect your Mattermost account to your Zoom account.
```

File: README.md
Comment (raghavaggarwal2308) [inline README.md:231]:
```suggestion
##### System admins only  
```

File: plugin.json
Comment (raghavaggarwal2308) [inline plugin.json:97]: It's not the bot who posts the meeting. The meeting is created by the user. We should update this wherever applicable.

**Review by raghavaggarwal2308 — CHANGES_REQUESTED** [review]

File: server/command.go
Comment (raghavaggarwal2308) [inline server/command.go:142]:
```suggestion
		return "Creating zoom meeting is disabled for this channel.", nil
```

File: server/command.go
Comment (raghavaggarwal2308) [inline server/command.go:292]:
```suggestion
			return "Updating zoom settings is disabled for this channel.", nil
```
Just to be consistant

File: server/command.go
Comment (raghavaggarwal2308) [inline server/command.go:308]: remove this

File: server/command.go
Comment (raghavaggarwal2308) [inline server/command.go:331]:
```suggestion
						HelpText:    "Enable to restrict creating meetings this channel.",
```

File: server/command.go
Comment (raghavaggarwal2308) [inline server/command.go:361]:
```suggestion
			return "Unable to open the dialog for setting preference", nil
```

File: server/command.go
Comment (raghavaggarwal2308) [inline server/command.go:380]: I think we should also show true/false in case of default

File: server/command.go
Comment (raghavaggarwal2308) [inline server/command.go:386]:
```suggestion
	return "Unknown Action", nil
```

File: server/command.go
Comment (raghavaggarwal2308) [inline server/command.go:426]:
```suggestion
	channelSettings := model.NewAutocompleteData("channel-settings", "", "Update current channel preference")
```

File: server/http.go
Comment (raghavaggarwal2308) [inline server/http.go:964]: Use "else if"

File: server/http.go
Comment (raghavaggarwal2308) [inline server/http.go:969]: If the channel type is private and no preference is set for it, we will be returning `false` in that case. Is this expected?

**Review by raghavaggarwal2308 — CHANGES_REQUESTED** [review]

File: server/http.go
Comment (ayusht2810) [reply server/http.go:967]: yes, by returning `false`, we are not restricting user from creating meeting links in private channel at line https://github.com/mattermost/mattermost-plugin-zoom/blob/6837578decc382b65d1acdb0f1541d6232cde73a/server/http.go#L510

File: README.md
Comment (raghavaggarwal2308) [inline README.md:207]:
```suggestion
* Enable the **Restrict Meeting Creation** configuration to restrict users from creating meetings in public channels```

File: README.md
Comment (raghavaggarwal2308) [inline README.md:233]:
```suggestion
- `/zoom channel-settings` :- To update channel preferences to restrict users from creating meetings in the current channel.
```

**Review by raghavaggarwal2308 — CHANGES_REQUESTED** [review]

**Review by raghavaggarwal2308 — COMMENTED** [review]
Can you add testcases for this new feature

File: server/http.go
Comment (raghavaggarwal2308) [inline server/http.go:943]: was this change necessary?

Comment (ayusht2810) [reply server/http.go:943]: Yes, `appErr` is of a different type, and the return type is different. Also, to get a proper error message, we are just returning the `Message` field from the `appErr`.

**Review by raghavaggarwal2308 — APPROVED** [review]

**Comment by raghavaggarwal2308** [comment]
@ayusht2810 Please resolve conflicts

**Comment by raghavaggarwal2308** [comment]
@mickmister Gentle reminder for review

**Comment by ayusht2810** [comment]
@mickmister Gentle reminder to review the PR

File: server/command.go
Comment (mickmister) [inline server/command.go:274]: Nit to use `p.client` instead of `p.API`. I wonder if we can make a golint rule for this. It doesn't make a huge difference, but it helps with references to `AppError` and `error` to avoid weird concrete vs interface things there. This `HasPermissionTo` method returns only a boolean so it's not any difference here

File: server/command.go
Comment (mickmister) [inline server/command.go:298]: This command doesn't have to do with specific channels does it? It's more of a "here's your user settings" thing right? Wondering why these changes are needed

File: server/command.go
Comment (mickmister) [inline server/command.go:308]: I'm thinking users that have permission to manage the channel can run `channel-settings`, but sysadmin privileges should be needed for `channel-settings list`

File: server/command.go
Comment (mickmister) [inline server/command.go:313]: Can we extract this block out into a `runEditChannelSettingsCommand` or something like that?

File: server/command.go
Comment (mickmister) [inline server/command.go:365]: and extract this into a `runListChannelSettingsCommand` function

File: server/configuration.go
Comment (mickmister) [inline server/configuration.go:45]: Probably good to add a comment to this setting too

```suggestion
	// RestrictMeetingCreation allows the admin to by default restrict zoom meetings by default to only private channels. The admin can also edit each channel's behavior with the `/zoom channel-settings` command
	RestrictMeetingCreation bool
```

File: server/http.go
Comment (mickmister) [inline server/http.go:559]: I know this manual JSON string pattern is already made throughout the project, but what do you think about this instead?

```go
type ErrorResponse struct {
    Error string `json:"error"`
}
```

```suggestion
		if err = json.NewEncoder(w).Encode(ErrorResponse{"Creating zoom meeting is disabled for this channel."}); err != nil {
```

File: server/http.go
Comment (mickmister) [inline server/http.go:578]: Similar thing with a `MeetingURLResponse` struct

File: server/http.go
Comment (mickmister) [inline server/http.go:692]: We're using the display name so probably best to put that here. A little unfortunate that this will become stale if the channel's display name changes. Maybe we should use `channel.Name` instead, which is generally more permanent? The user can change both values, but it's more likely that a display name gets changed. Storing it *with* the channel id below makes sense. I would expect the channel id to be in the struct as well, but it's obviously not required since the feature currently works without that

```suggestion
		ChannelDisplayName: submitRequest.CallbackId,
```

File: server/http.go
Comment (mickmister) [inline server/http.go:1006]: `DefaultPreference` seems a bit vague for a name for this value. Maybe `DefaultChannelRestrictionPreference`?

File: server/http.go
Comment (mickmister) [inline server/http.go:988]:
```suggestion
func (p *Plugin) isChannelRestrictedForMeetings(channelID string) (bool, int, error) {
```

Not sure why we need the http status code to be returned from here?

File: server/http.go
Comment (mickmister) [inline server/http.go:1025]: This needs to match a specific set of pre-defined values, so we should check that here

File: server/store.go
Comment (mickmister) [inline server/store.go:219]: I typically like to use `b` as the var name for `[]byte`. There exists a package `bytes` in the Go standard library, so this makes sure we don't shadow that library by accident

**Review by mickmister — APPROVED** [review]
LGTM :+1: just some non-blocking comments for discussion

---

For the markdown table in the `channel-settings list` response, we should probably have it say:

```
Default: Allow meetings in public channels, private channels, and DMs/GMs
```

or

```
Default: Allow meetings only in private channels and DMs/GMs
```

then simply omit any rows from the table that have the "default" value. We could probably just remove that channel from the saved settings map if the user chooses "default". 0/5 on this though

File: server/http.go
Comment (Kshitij-Katiyar) [reply server/http.go:988]: Returning `statusCode` for any error that arises so we can return the respective error.

File: server/command.go
Comment (Kshitij-Katiyar) [reply server/command.go:274]: We can make a seperate for adding golint rule for not using p.API since it will include a lot changes through the code once the new lint rule is implemented

**Comment by mickmister** [comment]
> @mickmister the default setting has a dynamic value (i.e. it can be updated in plugin configuration) Screenshot from 2024-06-06 18-06-05 Didn't get you on this, please explain.

@Kshitij-Katiyar Yes I was suggesting we display this value in the response in the wording I posted above. Just so when someone wants to inspect the settings on a channel, they can know what the default value is. Then we can omit any channels that are either unassigned or using the default value

**Comment by Kshitij-Katiyar** [comment]
> Default: Allow meetings only in private channels and DMs/GMs

this was the existing channel setting list 
channel setting zoom old

And this is the updated channel setting list
channel setting configuraiton new

@mickmister I hope this aligns with your suggestions. Please let us know if anything else is required.

**Comment by raghavaggarwal2308** [comment]
@fmartingr Gentle reminder for review

**Review by fmartingr — APPROVED** [review]
Apologies for the delay, just a nit about the help text.

File: server/command.go
Comment (ayusht2810) [reply server/command.go:332]: Yes, you got it right. If we enable it, it means we are restricting/disabling meeting creation in the current channel.

Comment (fmartingr) [reply server/command.go:332]: But then, the `enable` option below says: "Enable Zoom meetings in this channel". That was what seemed confusing to me.

File: server/command.go
Comment (mickmister) [inline server/command.go:300]:
```suggestion
	} else if params[0] == channelSettingsActionList {
```

**Comment by raghavaggarwal2308** [comment]
@mickmister Fixed the review comments added by you

File: server/command.go
Comment (raghavaggarwal2308) [reply server/command.go]: @fmartingr Please refer to the updated screenshots in the PR description. I hope its clear now.

Comment (fmartingr) [reply server/command.go]: Much more clear now, thank you!

**Review by mickmister — COMMENTED** [review]
LGTM, just one request to remove the two `bin` files

File: bin/golangci-lint
Comment (raghavaggarwal2308) [reply bin/golangci-lint:1]: @mickmister Removed

**Review by mickmister — APPROVED** [review]

**Review by AayushChaudhary0001 — COMMENTED** [review]
@ayusht2810 I found a issue while testing this PR.
Issue: Channel preferences are not shown for 'Default to plugin-wide settings (true)'

Steps to reproduce:

1. Set the channel preference to 'Default to plugin-wide settings(true)' for any channel
2. Now fetch the list of the channel preference of that channel

Actual: Channel preferences are not shown.
Expeceted: Accurate channel preferences are shown.

SS for the reference:
image (1)

**Comment by ayusht2810** [comment]
@AayushChaudhary0001 Fixed the above use case. Can you please re-test the PR?

**Review by AayushChaudhary0001 — APPROVED** [review]
The above change works well. The list is now visible to the users and all the functionalities related are working fine, LGTM. Approved.

**Comment by raghavaggarwal2308** [comment]
@wiggin77 Can you please review this PR. Its already approved but we need a codeowner's approval for merging

File: server/http.go
Comment (wiggin77) [inline server/http.go:553]:
```suggestion
		if err = json.NewEncoder(w).Encode(ErrorResponse{"Creating Zoom meeting is disabled for this channel."}); err != nil {
```

File: server/command.go
Comment (wiggin77) [inline server/command.go:146]: When "zoom" is used as a product name it should be capitalized.  This needs to be fixed for all cases.

**Review by wiggin77 — CHANGES_REQUESTED** [review]
Need to be consistent is capitalization for cases where the word "zoom" is used as a product name.

File: server/command.go
Comment (wiggin77) [inline server/command.go:17]: Capitalization missed here.

**Review by wiggin77 — CHANGES_REQUESTED** [review]
I see at least one missed capitalization.  Please do a full check of the source tree.

**Comment by raghavaggarwal2308** [comment]
@wiggin77 Updated. Please re-review

**Pull request merged**
